tests/src/test/scala/org/http4s/parser/Rfc3986ParserSpec.scala

```diff
@@ -0,0 +1,179 @@
+      val ipv6 = Ipv6Address(0, 0, 0, 0, 0, 1, 21787, 31704) // ::1:551b:7bd8
+      Rfc3986.ipv6.parse(ipv6.value).map(_._2) must beRight(ipv6)
+    }
+
+    "parse ipv6 address with zeros in the fourth and fifth spots" in {
+      val ipv6 = Ipv6Address(21494, -32768, 1, 0, 0, 1, 21787, 31704) // 53f6:8000:1::1:551b:7bd8
+      Rfc3986.ipv6.parse(ipv6.value).map(_._2) must beRight(ipv6)
+    }
+
+    "parse ipv6 address with zeros in the first six spots" in {
+      val ipv6 = Ipv6Address(0, 0, 0, 0, 0, 0, 21787, 31704) // ::551b:7bd8
+      Rfc3986.ipv6.parse(ipv6.value).map(_._2) must beRight(ipv6)
+    }
+
+    "parse ipv6 address with zeros in the fifth and sixth spots" in {
+      val ipv6 = Ipv6Address(21494, -32768, 1, 1, 0, 0, 21787, 31704) // 53f6:8000:1:1::551b:7bd8
+      Rfc3986.ipv6.parse(ipv6.value).map(_._2) must beRight(ipv6)
+    }
+
+    "parse ipv6 address with zeros in the first seven spots" in {
+      val ipv6 = Ipv6Address(0, 0, 0, 0, 0, 0, 0, 31704) // ::7bd8
+      Rfc3986.ipv6.parse(ipv6.value).map(_._2) must beRight(ipv6)
+    }
+
+    "parse ipv6 address with zeros in the sixth and seventh spots" in {
+      val ipv6 = Ipv6Address(21494, -32768, 1, 1, 21787, 0, 0, 31704) // 53f6:8000:1:1:551b::7bd8
+      Rfc3986.ipv6.parse(ipv6.value).map(_._2) must beRight(ipv6)
+    }
+
+    "parse ipv6 address with zeros in the all eight spots" in {
+      val ipv6 = Ipv6Address(0, 0, 0, 0, 0, 0, 0, 0) // ::
+      Rfc3986.ipv6.parse(ipv6.value).map(_._2) must beRight(ipv6)
+    }
+
+    "parse ipv6 address with zeros in the seventh and eighth spots" in {
+      val ipv6 = Ipv6Address(21494, -32768, 1, 1, 21787, 31704, 0, 0) // 53f6:8000:1:1:551b:7bd8::
+      Rfc3986.ipv6.parse(ipv6.value).map(_._2) must beRight(ipv6)
+    }
+
+    "parse ipv6 address with three zeros in a row" in {
+      val ipv6 = Ipv6Address(22322, -32768, 24194, 32767, 0, 0, 0, 1) // 5732:8000:5e82:7fff::1
+      Rfc3986.ipv6.parse(ipv6.value).map(_._2) must beRight(ipv6)
+    }
+
+    "not parse an invalid ipv6 address containing an invalid letter such as 'O'" in {
+      val invalidIp = "1200:0000:AB00:1234:O000:2552:7777:1313"
+      Rfc3986.ipv6.parse(invalidIp).map(_._2) must beLeft
+    }
+
+    "only parse part of an invalid ipv6 address where a single section is shortened (must be 2 or more)" in {
+      val invalidIp = "2001:db8::1:1:1:1:1"
+      Rfc3986.ipv6.parse(invalidIp).map(_._2.value) must beRight("2001:db8::1:1:1:1")
+    }
+
+    "only parse part of an ipv6 address where multiple sections are shortened (only 1 allowed)" in {
+      val invalidIp = "56FE::2159:5BBC::6594"
+      Rfc3986.ipv6.parse(invalidIp).map(_._2.value) must beRight("56fe::2159:5bbc")
+    }
+
+    "parse an ipv6 address which is too long (only parse what can be used)" in {
+      val ipv6 = "56FE:2159:5BBC:6594:1234:5678:9018:0123:2020"
+      Rfc3986.ipv6.parse(ipv6).map(_._2.value) must beRight(
+        "56fe:2159:5bbc:6594:1234:5678:9018:123")
+    }
```

Comment:
@lewisjkl Was there a particular reason to make a best effort on these instead of rejecting them as invalid?  I see some of these used to assert `Left`, and then changed around the time the `end` was dropped.  Are these cases we intentionally want to support, or just something that accidentally works now?

Reply:
The reason for this is because adding `end` into the parser made it so they couldn't compose with other parsers. For example, if you compose something like:

```scala
val myParser: Parser[String] = char('a').string <* end
val otherParser: Parser[String] = char('b').string

myParser ~ otherParser
```

I am pretty sure that this parser could never succeed since you can't have an `end` followed by more input. So I figured that the job of this `ipv6` parser would *not* be to reject extra input characters at the end of an ipv6 address, but rather that would have to be the job of a different parser that was aggregating this one as a part of its greater whole.

So come to think of it, OriginHeader should probably contain an `end` in its parser since we don't want anything to come after an ipv6/ipv4/host in that type of header. But it seems like ipv6/ipv4 should allow extra characters at the end of a valid address. What are your thoughts though?

Reply:
Oh, right.  These tests are calling `.parse`, not `.parseAll`.  So you're asserting that you're grabbing the part that's legal, and discarding the rest, which should then fail in `.parseAll`, or when composed with another parser that doesn't understand the leftovers.  I think this is correct.

I also don't think we generally want to call `.end`.  The caller uses `.parseAll` when it expect to consume the entire String.  But if we don't cap the parser with `.end`, we let a generic header parser look for optional trailing whitespace, or a headers (plural) parser parse the unconsumed input as the next header.  `.parseAll` lets the caller express "the end", when only the caller knows at what level they wish to parse.

Reply:
Gotcha yeah that makes sense, I didn't actually realize `parseAll` was a thing. Thanks!